codestyle/styles.md

```diff
@@ -0,0 +1,714 @@
+        background: #0F0;
+    }
+```
+
+**Indenting rulesets**
+
+* one (1) empty line between rulesets;
+* indent rulesets to signal their relation to each other in HTML;
+
+Example:
+
+```CSS
+.block {}
+
+    .block__element {}
+
+        .block__inner-element {}
+```
+
+* `::before` and `::after` pseudo-elements must be treated as inner elements and
+indented accordingly:
+
+```CSS
+.block {}
+
+    .block::before {}
+
+    .block::after {}
+```
+
+
+
+<a id="selectors"></a>
+## Selectors
+
+Keep selectors as short as possible, in order to keep specificity down and
+performance up.
+
+**Universal selector (`*`) is forbidden**
+
+```CSS
+/* Wrong: */
+
+* {
+    box-sizing: border-box;
+}
+```
+
+*Reasons:*
+
+* with high probability it will change rules that you were not going
+  to change;
+* because browsers read selectors right-to-left, the rightmost selector is often
+  critical in defining a selector’s performance;
+
+**No selectors by ID**
+
+```CSS
+/* Wrong: */
+
+#main {
+    color: #F00;
+}
+```
+
+*Reasons:*
+
+* selectors by id give too much weight for the rules under it;
+* it is impossible to use such blocks on the page more than one times;
+
+**No selectors by element**
+
+```CSS
+/* Wrong: */
+
+p {
+    margin: 1em 0;
+}
+
+.block img {
+    border: 1px solid #000;
+}
+```
+*Reasons:*
+
+* because with high probability the rules will be applied to elements you were
+  not going apply to;
+* inability to replace one tag on another without changing styles;
+
+*Exceptions:*
+
+* allowed to use in the reset/normalize styles files, which set basic
+  styles for the whole project;
+* allowed to use for styling selectors related to tables;
+* allowed to use for styling embedded a third-party widgets;
+
+**No selectors by attribute**
+
+```CSS
+/* Wrong: */
+
+.input[type="text"] {
+    font-size: 1em;
+}
+
+.input[disabled] {
+    opacity: .5;
+}
+
+
+/* Right: */
+
+.input_type_text {
+    font-size: 1em;
+}
+
+.input_state_disabled {
+    opacity: .5;
+}
+```
+
+*Reasons:*
+
+* attribute selectors has the same specificity as class selectors which
+  leads to an increase in the specificity and complexity of the redefinition of
+  the rules;
+
+*Exceptions:*
+
+* in case we need to style embedded a third-party widgets we have no ability to
+  edit its markup we can use this approach:
+
+```CSS
+[id="third-party-widget"] {}
+```
+
+
+**No cascade selectors**
+
+*Reasons:*
+
+- the cascade increases specificity and limits the ability to extend
+  and manipulate a codebase;
+
+*Exceptions:*
+
+- it's allowed to use a cascade if a block has a few themes and you need to add
+specific styles to the block elements, based on a theme. For example:
+
+```CSS
+.box {
+    background: #FFF;
+}
+
+    .box__title {
+        color: #000;
+    }
+
+.box_theme_dark {
+    background: #000;
+}
+
+    .box_theme_dark > .box__title {
+        color: #FFF;
+    }
+```
+
+*How to use:*
+
+- use no more than two class selectors at once;
```

Comment:
There's a case that I had to use 3 classes of selector:
For example:
.box_theme_dark.box_state_active > .box__title {
   color: #000;
}